From 607433e585409d1f31a20d3e6bfb51d625d52f70 Mon Sep 17 00:00:00 2001
From: Shivani Raval <sraval@habana.ai>
Date: Fri, 04 Apr 2025 09:58:29 +0300
Subject: [PATCH] [JSW-31] specs: added jgs files

Added similar files to FS1 for JGS asic build.
In future fs folder will be removed to gpu, which
has code for FS1 and JGS.

Change-Id: Idde4ee34d881d63cf786065815886dddd5890dd0
---

diff --git a/fs/jgs/address_space.h b/fs/jgs/address_space.h
new file mode 100644
index 0000000..4549bd5
--- /dev/null
+++ b/fs/jgs/address_space.h
@@ -0,0 +1,86 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2023 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+ #ifndef __FS1_ADDRESS_SPACE_H_
+ #define __FS1_ADDRESS_SPACE_H_
+
+ #define CFG_BAR_SIZE		0x10000000ull	/* 256MB */
+ #define CFG_PHYS_BASE		0x0300007FF0000000ull
+
+ #define HL_CFG_BASE			0x80000000
+ #define CFG_ARC_REGION_BASE		0xF0000000
+
+ #define IO0_PCIE_FW_SRAM_BASE		(mmIO_0_CFG_HALF_0_FW_MEM_PCIE_FW_SRAM_BASE +\
+					  mmPCIE_FW_SRAM_SRAM_LOCATIONS)
+ #define IO0_PCIE_FW_CSR_BASE		(mmIO_0_CFG_HALF_0_FW_MEM_PCIE_FW_SRAM_BASE +\
+					  mmPCIE_FW_SRAM_CSR_LOCATIONS)
+ #define IO0_FW_SRAM_BASE		mmIO_0_CFG_FW_MEM_FW_SRAM_BASE
+ #define IO0_SP_SRAM_BASE		mmIO_0_CFG_SP_RAM_BASE
+ #define IO0_RST_ICCM_BASE		mmIO_0_CFG_FW_MEM_RST_ARC_ICCM_BASE
+ #define IO0_RST_DCCM_BASE		mmIO_0_CFG_FW_MEM_RST_ARC_DCCM_BASE
+ #define IO0_MNG_ICCM_BASE		mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE
+ #define IO0_MNG_DCCM_BASE		mmIO_0_CFG_FW_MEM_MNG_ARC_DCCM_BASE
+ #define IO0_PID_ICCM_BASE		mmIO_0_CFG_FW_MEM_PID_ARC_ICCM_BASE
+ #define IO0_PID_DCCM_BASE		mmIO_0_CFG_FW_MEM_PID_ARC_DCCM_BASE
+ #define IO0_MISC_ICCM_BASE		mmIO_0_CFG_FW_MEM_MISC_ARC_ICCM_BASE
+ #define IO0_MISC_DCCM_BASE		mmIO_0_CFG_FW_MEM_MISC_ARC_DCCM_BASE
+ #define IO0_MUC_ICCM_BASE		mmIO_0_CFG_MEDIA_0_ICCM_BASE
+ #define IO0_MUC_DCCM_BASE		mmIO_0_CFG_MEDIA_0_DCCM_BASE
+
+ #define COMP0_RST_ICCM_BASE		mmCOMPUTE_0_CFG_FW_MEM_RST_ARC_ICCM_BASE
+ #define COMP0_RST_DCCM_BASE		mmCOMPUTE_0_CFG_FW_MEM_RST_ARC_DCCM_BASE
+ #define COMP0_MNG_ICCM_BASE		mmCOMPUTE_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE
+ #define COMP0_MNG_DCCM_BASE		mmCOMPUTE_0_CFG_FW_MEM_MNG_ARC_DCCM_BASE
+
+ #define LOCAL_ICCM_BASE			0
+ #define LOCAL_DCCM_BASE			0x80000000
+
+ /* 64b address used to convert specs 31bit address to full FS MEM address. Applicable to all dies */
+ #define FW_MEM_FULL_BASE		0x300007F80000000ull
+
+ #define IO0_PCIE_FW_SRAM_FULL_BASE	(FW_MEM_FULL_BASE + IO0_PCIE_FW_SRAM_BASE)
+ #define IO0_PCIE_FW_CSR_FULL_BASE	(FW_MEM_FULL_BASE + IO0_PCIE_FW_CSR_BASE)
+ #define IO0_FW_SRAM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_FW_SRAM_BASE)
+ #define IO0_SP_SRAM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_SP_SRAM_BASE)
+ #define IO0_RST_ICCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_RST_ICCM_BASE)
+ #define IO0_RST_DCCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_RST_DCCM_BASE)
+ #define IO0_MNG_ICCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_MNG_ICCM_BASE)
+ #define IO0_MNG_DCCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_MNG_DCCM_BASE)
+ #define IO0_PID_ICCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_PID_ICCM_BASE)
+ #define IO0_PID_DCCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_PID_DCCM_BASE)
+ #define IO0_MISC_ICCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_MISC_ICCM_BASE)
+ #define IO0_MISC_DCCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_MISC_DCCM_BASE)
+ #define IO0_MUC_ICCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_MUC_ICCM_BASE)
+ #define IO0_MUC_DCCM_FULL_BASE		(FW_MEM_FULL_BASE + IO0_MUC_DCCM_BASE)
+
+ #define COMP0_RST_ICCM_FULL_BASE	(FW_MEM_FULL_BASE + COMP0_RST_ICCM_BASE)
+ #define COMP0_RST_DCCM_FULL_BASE	(FW_MEM_FULL_BASE + COMP0_RST_DCCM_BASE)
+ #define COMP0_MNG_ICCM_FULL_BASE	(FW_MEM_FULL_BASE + COMP0_MNG_ICCM_BASE)
+ #define COMP0_MNG_DCCM_FULL_BASE	(FW_MEM_FULL_BASE + COMP0_MNG_DCCM_BASE)
+
+ #define IO_FW_SRAM_SIZE			SZ_128K
+ #define IO_SP_SRAM_SIZE			SZ_1M
+ #define IO_RST_ICCM_SIZE		SZ_64K
+ #define IO_RST_DCCM_SIZE		SZ_32K
+ #define IO_MNG_ICCM_SIZE		SZ_256K
+ #define IO_MNG_DCCM_SIZE		SZ_128K
+ #define IO_PID_ICCM_SIZE		SZ_128K
+ #define IO_PID_DCCM_SIZE		SZ_64K
+ #define IO_MISC_ICCM_SIZE		SZ_256K
+ #define IO_MISC_DCCM_SIZE		SZ_64K
+ #define IO_MUC_ICCM_SIZE		SZ_128K
+ #define IO_MUC_DCCM_SIZE		SZ_128K
+
+ #define COMP_RST_ICCM_SIZE		SZ_64K
+ #define COMP_RST_DCCM_SIZE		SZ_32K
+ #define COMP_MNG_ICCM_SIZE		SZ_256K
+ #define COMP_MNG_DCCM_SIZE		SZ_128K
+
+ #define ADDR_MASK(_mem_type)		(_mem_type ## _SIZE - 1)
+
+ #endif /* __FS1_ADDRESS_SPACE_H_ */
+
diff --git a/fs/jgs/boot_rom_export_dev.h b/fs/jgs/boot_rom_export_dev.h
new file mode 100644
index 0000000..8831490
--- /dev/null
+++ b/fs/jgs/boot_rom_export_dev.h
@@ -0,0 +1,62 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+#ifndef SPECS_JGS_BOOT_ROM_EXPORT_DEV_H_
+#define SPECS_JGS_BOOT_ROM_EXPORT_DEV_H_
+
+struct mem_copy_flag {
+	uint8_t copy_type;
+	uint8_t mem_type;
+};
+
+enum dma_error {
+	DMA_SUCCESS = 0,
+	DMA_INVALID_CRC_ERROR,
+	DMA_INVALID_DEST_TYPE,
+	DMA_UNALIGNED_IMG_SIZE,
+	DMA_UNALIGNED_DEST_ADDR,
+	DMA_INVALID_IMG_SIZE,
+	DMA_ERROR_MAX
+};
+
+int boot_rom_mem_init(uint64_t addr, uint32_t size) __attribute__ ((long_call));
+
+int boot_rom_mem_copy(struct mem_copy_flag flag, uint64_t src, uint64_t dst,
+			uint32_t size)__attribute__ ((long_call));
+
+uint32_t boot_rom_fw_mem_arc_addr(uint64_t addr) __attribute__ ((long_call));
+
+/**
+ * @brief perform spi dma
+ *
+ * This API does spi dma from spi to fw mem or hbw memory
+ *
+ * @param image_size size of the image to be dma from spi
+ * @param flash_offset start offset of image in flash
+ * @param dest_addr fw mem or hbw address where image should be copied
+ * @param skip_crc_check api will skip crc check when true
+ * @param manual_crc api will perform manual crc check at the end of dma when true
+ * @param expected_crc 32-bit crc value to compare with at the end of dma. Valid
+ * only if manual_crc is true.
+ */
+enum dma_error boot_rom_spi_dma(uint32_t image_size, uint32_t flash_offset,
+				uint64_t dest_addr, bool skip_crc_check,
+				bool manual_crc, uint32_t expected_crc)__attribute__ ((long_call));
+
+int secure_boot_verify_image_hash384(uint8_t *exp_hash, unsigned char *img,
+				uint32_t img_len) __attribute__ ((long_call));
+
+int lfs_file_set(const char *file_tag, uint32_t type, const void *data, uint32_t size)
+		__attribute__ ((long_call));
+
+int lfs_file_copy(const char *file_tag, uint32_t type, void *desc_data, uint32_t size)
+		__attribute__ ((long_call));
+
+int lfs_file_copy_partial(const char *file_tag, uint32_t type, void *data,
+			uint32_t offset, uint32_t size) __attribute__ ((long_call));
+
+uint32_t lfs_disk_set_empty(void) __attribute__ ((long_call));
+
+#endif /* SPECS_FS1_BOOT_ROM_EXPORT_DEV_H_ */
diff --git a/fs/jgs/jgs_boot_info.h b/fs/jgs/jgs_boot_info.h
new file mode 100644
index 0000000..5baf067
--- /dev/null
+++ b/fs/jgs/jgs_boot_info.h
@@ -0,0 +1,41 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2024 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+ #ifndef JGS_BOOT_INFO_H
+ #define JGS_BOOT_INFO_H
+
+ #define HL_BOOT_INFO_VERSION	0x1
+
+ /*
+  * Descriptor version change log:
+  * 1 -> Initial version
+  *      hl_fw_boot_ptrs:
+  *              > boot_ptrs_img_sz : Size of the complete manifest to dma.
+  *              > img_desc_sz : Size of a single image descriptor.
+  *              > hl_fw_boot_magic : Packed with ['N','X','T', hl_fw_boot_ptr_imgs]
+  *              > count : Image descriptors count
+  *              > pub_key_sz : public key size
+  *              > sig_sz : signature size
+  *              > sec_ver : security version
+  *              > key_idx : key index to be validated
+  *              > offset2array : offset to array of offsets
+  *      hl_fw_img_desc:
+  *              > hl_fw_boot_magic : Packed with ['P','T','R', hl_fw_img_type]
+  *              > desc_ver : descriptor version
+  *              > idx : index used to identify primary and secondary image
+  *              > offset[2] : Image offset on SPI
+  *              > size[2] : Image size array
+  *              > load_high[2] : high [32-63] bits of the load address
+  *              > load_low[2] : low [0-31] bits of the load address
+  *              > chksum[2] : sha384 checksum
+  *              > start_high[2] : high [32-63] bits of the start address
+  *              > start_low[2] : low [0-31] bits of the start address
+  */
+
+ #define HL_FW_IMG_DESC_CUR_VER	1
+
+ #endif
diff --git a/fs/jgs/jgs_efuse.h b/fs/jgs/jgs_efuse.h
new file mode 100644
index 0000000..14437b5
--- /dev/null
+++ b/fs/jgs/jgs_efuse.h
@@ -0,0 +1,121 @@
+/* SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright 2024 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+ #ifndef _SPECS_JGS_EFUSE_H_
+ #define _SPECS_JGS_EFUSE_H_
+
+ #define IO_EFUSE_SYSCFG_OFFSET		60
+ #define IO_EFUSE_SYSCFG2_OFFSET		61
+ #define IO_EFUSE_SVN_FW_OFFSET		63
+
+ #define IO_EFUSE_SVN_FW_DWORD_COUNT	4
+
+ #define EFUSE_WORDS_PER_BANK		128
+
+ /* runtime settings for PALAD use only (using spare address in IO fuse) */
+ #define IO_EFUSE_PALAD_CFG_0_OFFSET	47
+ #define SRIOV_NUM_OF_VFS_M		GENMASK(2, 0) /* number of SRIOV virtual functions */
+ #define ATS_CAPABILITY_ENABLED_M	BIT(3)	/* enable PCI ATS capability */
+ #define CXL_CACHE_CAP_ENABLED_M		BIT(4)	/* enable CXL.Cache capability */
+ #define BLOADER_COPY_LARGE_APP		BIT(5)	/* app is bigger than single ICCM memory */
+ #define BLOADER_COPY_BOOTINFO		BIT(6)	/* init bootinfo section in spsram from dccm */
+ #define BMU_INIT_DEVEL_MODE		BIT(7)	/* init BMU development mode configuration */
+ #define BMU_INIT_PROD_MODE		BIT(8)	/* init BMU production mode configuration */
+ #define CRS_MODE_ENABLED_M		BIT(9) 	/* enable CRS mode */
+ #define SIM_MODE			BIT(10)	/* Simulator mode */
+
+ /* runtime settings for PALAD use only (using spare address in all efuses) */
+ #define EFUSE_PALAD_AVAILABLE_DIES_OFFSET		46
+ #define PALAD_AVAILABLE_DIES_M				GENMASK(7, 0)
+ #define COMPUTE_DIE_0_AVAIL_M				BIT(0)
+ #define COMPUTE_DIE_1_AVAIL_M				BIT(1)
+ #define COMPUTE_DIE_2_AVAIL_M				BIT(2)
+ #define COMPUTE_DIE_3_AVAIL_M				BIT(3)
+ #define IO_DIE_0_AVAIL_M				BIT(4)
+ #define IO_DIE_0_HALF_1_AVAIL_M				BIT(5)
+ #define IO_DIE_1_AVAIL_M				BIT(6)
+ #define IO_DIE_1_HALF_1_AVAIL_M				BIT(7)
+
+ /* EFUSE SYSCFG BIT Definitions start */
+ #define EFUSE_SYSCFG_REVOKE_KEY_0			BIT(0)
+ #define EFUSE_SYSCFG_REVOKE_KEY_0_1			(BIT(0) | BIT(1))
+ #define EFUSE_SYSCFG_REVOKE_KEY_0_1_2			(BIT(0) | BIT(1) | BIT(2))
+ #define EFUSE_SYSCFG_REVOKE_KEY_0_1_2_3			(BIT(0) | BIT(1) | BIT(2) | BIT(3))
+ #define EFUSE_SYSCFG_REVOKE_KEY_4			BIT(4)
+ #define EFUSE_SYSCFG_KEY_REVOCATION_BITS		((EFUSE_SYSCFG_REVOKE_KEY_0_1_2_3) \
+							 | (EFUSE_SYSCFG_REVOKE_KEY_4))
+ #define EFUSE_SYSCFG_EOM_BIT				BIT(5)
+ #define EFUSE_SYSCFG_DIS_IB_REC_FW_UPDATE		BIT(6)
+ #define EFUSE_SYSCFG_EN_DIE_ID_SECURITY_CHK		BIT(7)
+ #define EFUSE_SYSCFG_EN_SECURE_BOOT			BIT(9)
+ #define EFUSE_SYSCFG_SIG_VAL_EN				BIT(10)
+ #define EFUSE_SYSCFG_DIS_TWI				BIT(11)
+ #define EFUSE_SYSCFG_DIS_JTAG				BIT(12)
+ #define EFUSE_SYSCFG_DIS_EFUSE_OVERRIDE			BIT(13)
+ #define EFUSE_SYSCFG_VOLT_PROT_BITS			(BIT(14) | BIT(15) | BIT(16) | BIT(17))
+ #define EFUSE_SYSCFG_PCIE_GEN_LIMIT			BIT(18)
+ #define EFUSE_SYSCFG_SPI_DIRECT_FW_UPDATE		BIT(19)
+ #define EFUSE_SYSCFG_BLOCK_ALL_RR			BIT(20)
+ #define EFUSE_SYSCFG_DIS_SECONDARY_IMG			BIT(21)
+ #define EFUSE_SYSCFG_SPI_TPM_BAUDR			BIT(22)
+ #define EFUSE_SYSCFG_SKU				(BIT(23) | BIT(24) | BIT(25) | BIT(26))
+ #define EFUSE_SYSCFG_PSOC_SPI_GW_BAUDR			BIT(27)
+ #define EFUSE_SYSCFG_DIS_PSOC_SPI_GW			BIT(28)
+ #define EFUSE_SYSCFG_SPI_TPM_RX_DELAY			BIT(29)
+ #define EFUSE_SYSCFG_GPIO01_RECOVERY_CONTROL		BIT(30)
+ #define EFUSE_SYSCFG_EROM_I3C_INIT			BIT(31)
+ /* EFUSE SYSCFG BIT Definitions end */
+
+ /* EFUSE SYSCFG 2 BIT Definitions start */
+ #define EFUSE_SYSCFG2_RECOVERY_CODE_LOCATION		BIT(27)
+ #define EFUSE_SYSCFG2_DISABLE_BOOT_INFO			BIT(28)
+ #define EFUSE_SYSCFG2_RMA_MODE				BIT(29)
+ /* EFUSE SYSCFG 2 BIT Definitions end */
+
+ /*SPI TPM BAUDR configurations*/
+ enum boot_rom_tpm_spi_freq_config {
+	 FREQ_CFG_12_5MHZ = 0,
+	 FREQ_CFG_6_25MHZ = 1,
+ };
+
+ /* EFUSE SYSCFG2 BIT Definitions start */
+ #define EFUSE_SYSCFG2_D2D_DISABLE			BIT(31)
+ /* EFUSE SYSCFG2 BIT Definitions end */
+
+ /* EFUSE Shift definitions */
+ #define EFUSE_SYSCFG_SKU_SHIFT				23
+
+ #define EFUSE_SYSCFG_PRE_PRODUCTION_FLAGS		((EFUSE_SYSCFG_EN_SECURE_BOOT \
+							 | EFUSE_SYSCFG_DIS_EFUSE_OVERRIDE \
+							 | EFUSE_SYSCFG_SIG_VAL_EN \
+							 | EFUSE_SYSCFG_REVOKE_KEY_0 \
+							 | EFUSE_SYSCFG_DIS_TWI /* TODO: TBD */ \
+							 | EFUSE_SYSCFG_DIS_JTAG)) /* TODO: TBD */
+
+ /* redundnacy value read from EFUSE start */
+ #define EFUSE_IO_VDEC_AND_NIC_REDUN_OFFSET		19
+ #define EFUSE_IO_VDEC_REDUN_MASK			GENMASK(17, 0)
+ #define EFUSE_IO_NIC_REDUN_MASK				GENMASK(31, 19)
+ #define IO_VDEC8_FAULT_MASK_D0_L			GENMASK(8, 0)
+ #define IO_VDEC8_FAULT_MASK_D0_H			GENMASK(17, 9)
+ #define IO_DEC8_FAULT_MASK_DEFAULT			BIT(8)
+ #define IO_NIC12_FAULT_MASK_DEFAULT			BIT(12)
+ #define IO_VDEC8_NUM_PER_HALF_DIE			9
+
+ #define EFUSE_COMPUTE_EDMA_REDN_OFFSET			29
+ #define EFUSE_COMPUTE_EDMA_REDN_MASK			GENMASK(4, 0)
+
+ #define EFUSE_COMPUTE_PDMA_REDN_OFFSET			29
+ #define EFUSE_COMPUTE_PDMA_REDN_MASK			GENMASK(6, 5)
+
+ #define EFUSE_COMPUTE_XECORE_0_REDUN_OFFSET		24
+ #define EFUSE_COMPUTE_XECORE_1_REDUN_OFFSET		26
+
+ #define COMPUTE_EDMA_3_4_FAULT_MASK_DEFAULT		GENMASK(4, 3)
+ #define COMPUTE_PDMA_2_FAULT_MASK_DEFAULT		BIT(2)
+ #define COMPUTE_XECORE_17_FAULT_MASK_DEFAULT		BIT(16) /* TODO xecore number confirm */
+
+ #endif /* _SPECS_FS1_EFUSE_H_ */
diff --git a/fs/jgs/jgs_fw_if.h b/fs/jgs/jgs_fw_if.h
new file mode 100644
index 0000000..230bcf7
--- /dev/null
+++ b/fs/jgs/jgs_fw_if.h
@@ -0,0 +1,151 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2019-2023 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+ #ifndef JGS_FW_IF_H
+ #define JGS_FW_IF_H
+
+ /* rst src bit in PSOC_GLOBAL_CONF_RST_SRC register */
+ enum fs1_rst_src_bit {
+	 HL_COLD_RST,
+	 HL_MANUAL_RST,
+	 HL_PRSTN_RST,
+	 HL_SOFT_RST,
+	 HL_WD_RST,
+	 HL_FW_ALL_RST,
+	 HL_SW_ALL_RST,
+	 HL_FLR_RST,
+	 HL_ECC_DERR_RST,
+	 HL_CXL_RST,
+	 HL_HOT_RST,
+	 HL_VF_FLR_RST,
+	 HL_UNKNOWN_RST = 255
+ };
+
+ /*
+  * CPU ID must be written before the CPU reset is released. Since ARC MNG is the first
+  * to run (by HW), its ID will be the default (0). For the other CPUs a unique ID
+  * is set.
+  */
+ #define IO0_PSOC_ARC_MNG_ID		0
+ #define IO0_PSOC_ARC_RST_ID		1
+ #define IO0_PSOC_ARC_PID_ID		2
+ #define IO0_PSOC_ARC_MISC_ID		3
+ #define IO0_PCIE_ARC_CXL_ID		4
+
+ struct fs1_cold_rst_data {
+	 union {
+		 struct {
+			 uint32_t recovery_flag: 1;
+			 uint32_t validation_flag: 1;
+			 uint32_t valid_manifest_idx: 1;
+			 uint32_t efuse_read_flag: 1;
+			 uint32_t spsram_init_done : 1;
+			 uint32_t arc1_iccm_init_done: 1;
+			 uint32_t tpm_validation_flag: 1;
+			 uint32_t arc1_persist_dccm_init_done: 1;
+			 uint32_t fuse_emulation_init_done: 1;
+			 uint32_t boot_info_set_done: 1;
+			 uint32_t reserved : 22;
+		 };
+		 uint32_t data;
+	 };
+ };
+
+ #define FS1_NUM_HALF_DIES		2
+ #define FS1_NUM_IO_DIES		2
+ #define FS1_NUM_COMPUTE_DIES		4
+ #define FS1_NUM_REDUNDANT_EDMA		2
+
+ #define MAX_DIE_COUNT		(FS1_NUM_IO_DIES + FS1_NUM_COMPUTE_DIES)
+
+ /* In case there is a defect, io die redundant components are
+  * 1 of 9 DEC in each IO Die half
+  * 1 of 13 NICs in each IO Die
+  * PCIE_CLS in IO Die 1
+  * PMMU in IO Die for production (only used in IO Die stand-alone package)
+  * redundancy structure for FS1 IO die
+  */
+ struct fs1_redundancy_ctx_io {
+	 uint8_t initialized;		/* track driver initialization */
+	 uint16_t dec_fault_mask[FS1_NUM_HALF_DIES];
+	 uint16_t nic_fault_mask;
+	 uint8_t pmmu_fault_mask;
+	 uint8_t pcie_cls_fault_mask;	/* used only for die 1 */
+
+	 int8_t redundant_dec[FS1_NUM_HALF_DIES];
+	 int8_t redundant_nic;
+	 int8_t redundant_pmmu;
+	 int8_t redundant_die_pcie_cls;	/* used only for die 1 */
+ };
+
+ /* compute die redundant components are
+  * 2 of 5 EDMAs in each Compute Die
+  * 1 of 3 PDMAs in each Compute Die
+  * 1 of every 17 XeCores in Compute Die
+  * 1 CS + 6x HBM PC , AKA Dual plane.
+  * redundancy structure for FS1 compute die
+  */
+ struct fs1_redundancy_ctx_compute {
+	 uint8_t initialized;		/* track driver initialization */
+	 uint8_t edma_fault_mask;
+	 uint8_t pdma_fault_mask;
+	 uint32_t xecore_fault_mask;
+
+	 /* TODO  Dual Plane */
+	 int8_t redundant_edma[FS1_NUM_REDUNDANT_EDMA];
+	 int8_t redundant_pdma;
+	 int8_t redundant_xecore;
+ };
+
+ /* redundancy structure for FS1 */
+ struct fs1_redundancy_ctx {
+	 struct fs1_redundancy_ctx_io io_die[FS1_NUM_IO_DIES];
+	 struct fs1_redundancy_ctx_compute compute_die[FS1_NUM_COMPUTE_DIES];
+ };
+
+ /*
+  * Note:
+  * IO Half1 Die ID's are Virtual, just used for register offsets.
+  * Actual die ids are same between half0 and half1.
+  */
+ enum fs1_die_id {
+	 FS1_COMPUTE_DIE_0 = 0,
+	 FS1_COMPUTE_DIE_1 = 1,
+	 FS1_COMPUTE_DIE_2 = 2,
+	 FS1_COMPUTE_DIE_3 = 3,
+	 FS1_IO_DIE_0 = 4,
+	 FS1_IO_DIE_0_HALF_1 = 5,	/* Virtual Die ID */
+	 FS1_IO_DIE_1 = 6,
+	 FS1_IO_DIE_1_HALF_1 = 7,	/* Virtual Die ID */
+	 FS1_INVALID_DIE_ID = 8,
+ };
+
+ enum fs1_die_type {
+	 FS1_COMPUTE_DIE,
+	 FS1_IO_DIE_MASTER,
+	 FS1_IO_DIE_SLAVE,
+	 FS1_INVALID_DIE_TYPE,
+ };
+
+ enum fs1_sku_type {
+	 FS1_FULL_PACKAGE,
+	 FS1_PCIE_CARD,
+	 FS1_IO_DIE_SA,
+	 FS1_INVALID_SKU,
+ };
+
+ /* This enum used as index while fake fuse bin creation and to retrieve */
+ enum fake_fuse_bin_index {
+  FAKE_FUSE_IO_D0,
+  FAKE_FUSE_IO_D1,
+  FAKE_FUSE_COMPUTE_D0,
+  FAKE_FUSE_COMPUTE_D1,
+  FAKE_FUSE_COMPUTE_D2,
+  FAKE_FUSE_COMPUTE_D3,
+ };
+
+ #endif /* JGS_FW_IF_H */
diff --git a/fs/jgs/jgs_mem_addr_dt.h b/fs/jgs/jgs_mem_addr_dt.h
new file mode 100644
index 0000000..2d187fd
--- /dev/null
+++ b/fs/jgs/jgs_mem_addr_dt.h
@@ -0,0 +1,243 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2023 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef __JGS_MEM_ADDR_H_
+#define __JGS_MEM_ADDR_H_
+
+/*
+ * Notes:
+ * 1. The following definitions are used by DT (device tree).
+ * 2. The missing "0x" prefix in <>_section definitions is added in DT.
+ * 3. <>_section postfix means that the specific section is not only used by DT but also
+ *      in post-build.
+ * 4. <>_offset and <>_size matches memory section offset and size in DT and post-build.
+ *      The <>_offset is in arc view.
+ * 5. <>_hi and <>_low is section offset in AXI view.
+ */
+
+/*
+ * ToDo: Remove these macros after successful merge of
+ *       io die and compute die macros for preboot.
+*/
+
+#define ARC_PREBOOT_ICCM_OFFSET	0
+#define ARC_PREBOOT_ICCM_SIZE	0x100
+#define ARC_PREBOOT_DCCM_OFFSET	80000000
+#define ARC_PREBOOT_DCCM_SIZE	0x80
+
+#define ARC_PREBOOT_IO_ICCM_OFFSET	0
+#define ARC_PREBOOT_IO_ICCM_SIZE	0x100
+#define ARC_PREBOOT_IO_DCCM_OFFSET	80000000
+#define ARC_PREBOOT_IO_DCCM_SIZE	0x80
+
+#define ARC_PREBOOT_COMPUTE_ICCM_OFFSET	0
+#define ARC_PREBOOT_COMPUTE_ICCM_SIZE	0x80
+#define ARC_PREBOOT_COMPUTE_DCCM_OFFSET	80000000
+#define ARC_PREBOOT_COMPUTE_DCCM_SIZE	0x40
+
+#define SPSRAM_PREBOOT_OFFSET	78a00000
+#define SPSRAM_PREBOOT_SIZE	0x400
+#define BOOTINFO_REGION_OFFSET	00000
+#define BOOTINFO_REGION_SIZE	0x20
+
+#define ARC_DSM_ICCM_OFFSET	0
+#define ARC_DSM_ICCM_SIZE	0x100
+#define ARC_DSM_DCCM_OFFSET	80000000
+#define ARC_DSM_DCCM_SIZE	0x100
+
+/* TODO: remove _RST_ macros */
+#define ARC_RST_ICCM_OFFSET	0
+#define ARC_RST_ICCM_SIZE	0x40
+#define ARC_RST_DCCM_OFFSET	80000000
+#define ARC_RST_DCCM_SIZE	0x20
+
+#define ARC_RESET_ICCM_SECTION	1
+#define ARC_RESET_DCCM_SECTION	1
+#define ARC_RESET_ICCM_OFFSET	0
+#define ARC_RESET_ICCM_SIZE	0x40
+#define ARC_RESET_DCCM_OFFSET	80000000
+#define ARC_RESET_DCCM_SIZE	0x20
+
+#define ARC_MGMT_ICCM_SECTION	1
+#define ARC_MGMT_ICCM_OFFSET	0
+#define ARC_MGMT_ICCM_HI	0x0300007F
+#define ARC_MGMT_ICCM_LO	0xF8840000
+#define ARC_MGMT_ICCM_SIZE	0x100
+#define ARC_MGMT_DCCM_SECTION	1
+#define ARC_MGMT_DCCM_OFFSET	80000000
+#define ARC_MGMT_DCCM_HI	0x0300007F
+#define ARC_MGMT_DCCM_LO	0xF8880000
+#define ARC_MGMT_DCCM_SIZE	0x80
+#define ARC_MGMT_CPU_HI		0x0300007F
+#define ARC_MGMT_CPU_LO		0xF9CE0000
+
+#define ARC_MGMT_COMPUTE_ICCM_SECTION	1
+#define ARC_MGMT_COMPUTE_ICCM_OFFSET	0
+#define ARC_MGMT_COMPUTE_ICCM_HI	0x0300007F
+#define ARC_MGMT_COMPUTE_ICCM_LO	0xF0840000
+#define ARC_MGMT_COMPUTE_ICCM_SIZE	0x80
+#define ARC_MGMT_COMPUTE_DCCM_SECTION	1
+#define ARC_MGMT_COMPUTE_DCCM_OFFSET	80000000
+#define ARC_MGMT_COMPUTE_DCCM_HI	0x0300007F
+#define ARC_MGMT_COMPUTE_DCCM_LO	0xF0880000
+#define ARC_MGMT_COMPUTE_DCCM_SIZE	0x40
+
+#define ARC_PID_ICCM_SECTION	1
+#define ARC_PID_ICCM_OFFSET	0
+#define ARC_PID_ICCM_HI		0x0300007F
+#define ARC_PID_ICCM_LO		0xF88C0000
+#define ARC_PID_ICCM_SIZE	0x80
+#define ARC_PID_DCCM_SECTION	1
+#define ARC_PID_DCCM_OFFSET	80000000
+#define ARC_PID_DCCM_HI		0x0300007F
+#define ARC_PID_DCCM_LO		0xF88E0000
+#define ARC_PID_DCCM_SIZE	0x40
+#define ARC_PID_CPU_HI		0x0300007F
+#define ARC_PID_CPU_LO		0xF9CE8000
+
+#define SPI_FLASH_REGION_IMAGE_DESC_PRIMARY_OFFSET		0
+#define SPI_FLASH_REGION_IMAGE_DESC_PRIMARY_SIZE		10000
+#define SPI_FLASH_REGION_PPBOOT_PRIMARY_OFFSET			10000
+#define SPI_FLASH_REGION_PPBOOT_PRIMARY_SIZE			20000
+#define SPI_FLASH_REGION_PREBOOT_IO_DIE_PRIMARY_OFFSET		30000
+#define SPI_FLASH_REGION_PREBOOT_IO_DIE_PRIMARY_SIZE		40000
+#define SPI_FLASH_REGION_PREBOOT_COMPUTE_DIE_PRIMARY_OFFSET	70000
+#define SPI_FLASH_REGION_PREBOOT_COMPUTE_DIE_PRIMARY_SIZE	20000
+#define SPI_FLASH_REGION_CXL_ARC_ICCM_PRIMARY_OFFSET		90000
+#define SPI_FLASH_REGION_CXL_ARC_ICCM_PRIMARY_SIZE		20000
+#define SPI_FLASH_REGION_CXL_ARC_DCCM_PRIMARY_OFFSET		b0000
+#define SPI_FLASH_REGION_CXL_ARC_DCCM_PRIMARY_SIZE		20000
+#define SPI_FLASH_REGION_DSM_ARC_ICCM_PRIMARY_OFFSET		d0000
+#define SPI_FLASH_REGION_DSM_ARC_ICCM_PRIMARY_SIZE		20000
+#define SPI_FLASH_REGION_DSM_ARC_DCCM_PRIMARY_OFFSET		f0000
+#define SPI_FLASH_REGION_DSM_ARC_DCCM_PRIMARY_SIZE		20000
+#define SPI_FLASH_REGION_RECOVERY_IMAGE_PCIE_PRIMARY_OFFSET	110000
+#define SPI_FLASH_REGION_RECOVERY_IMAGE_PCIE_PRIMARY_SIZE	10000
+#define SPI_FLASH_REGION_RECOVERY_IMAGE_FW_UPD_PRIMARY_OFFSET	120000
+#define SPI_FLASH_REGION_RECOVERY_IMAGE_FW_UPD_PRIMARY_SIZE	20000
+#define SPI_FLASH_REGION_FAKE_FUZE_PRIMARY_OFFSET		140000
+#define SPI_FLASH_REGION_FAKE_FUZE_PRIMARY_SIZE			10000
+#define SPI_FLASH_REGION_RESERVED_1_OFFSET			150000
+#define SPI_FLASH_REGION_RESERVED_1_SIZE			b0000
+#define SPI_FLASH_REGION_IMAGE_DESC_SECONDARY_OFFSET		200000
+#define SPI_FLASH_REGION_IMAGE_DESC_SECONDARY_SIZE		10000
+#define SPI_FLASH_REGION_PPBOOT_SECONDARY_OFFSET		210000
+#define SPI_FLASH_REGION_PPBOOT_SECONDARY_SIZE			20000
+#define SPI_FLASH_REGION_PREBOOT_IO_DIE_SECONDARY_OFFSET	230000
+#define SPI_FLASH_REGION_PREBOOT_IO_DIE_SECONDARY_SIZE		40000
+#define SPI_FLASH_REGION_PREBOOT_COMPUTE_DIE_SECONDARY_OFFSET	270000
+#define SPI_FLASH_REGION_PREBOOT_COMPUTE_DIE_SECONDARY_SIZE	20000
+#define SPI_FLASH_REGION_CXL_ARC_ICCM_SECONDARY_OFFSET		290000
+#define SPI_FLASH_REGION_CXL_ARC_ICCM_SECONDARY_SIZE		20000
+#define SPI_FLASH_REGION_CXL_ARC_DCCM_SECONDARY_OFFSET		2b0000
+#define SPI_FLASH_REGION_CXL_ARC_DCCM_SECONDARY_SIZE		20000
+#define SPI_FLASH_REGION_DSM_ARC_ICCM_SECONDARY_OFFSET		2d0000
+#define SPI_FLASH_REGION_DSM_ARC_ICCM_SECONDARY_SIZE		20000
+#define SPI_FLASH_REGION_DSM_ARC_DCCM_SECONDARY_OFFSET		2f0000
+#define SPI_FLASH_REGION_DSM_ARC_DCCM_SECONDARY_SIZE		20000
+#define SPI_FLASH_REGION_RECOVERY_IMAGE_PCIE_SECONDARY_OFFSET	310000
+#define SPI_FLASH_REGION_RECOVERY_IMAGE_PCIE_SECONDARY_SIZE	10000
+#define SPI_FLASH_REGION_RECOVERY_IMAGE_FW_UPD_SECONDARY_OFFSET	320000
+#define SPI_FLASH_REGION_RECOVERY_IMAGE_FW_UPD_SECONDARY_SIZE	20000
+#define SPI_FLASH_REGION_FAKE_FUZE_SECONDARY_OFFSET		340000
+#define SPI_FLASH_REGION_FAKE_FUZE_SECONDARY_SIZE		10000
+#define SPI_FLASH_REGION_RESERVED_2_OFFSET			350000
+#define SPI_FLASH_REGION_RESERVED_2_SIZE			b0000
+
+/*
+ * FS1 Dup engine supports two regions with 8 groups each. Region 0
+ * On IO die, Region 0 is allocated to mgmt arc, Region 1 is
+ * allocated to pid Arc.
+ * On Compute die, Region 0 is allocated to mgmt arc, Region 1 is
+ * not used
+ * DUP_GROUP_MMIO_RANGE = 512, DUP_NUM_GROUPS_PER_REGION =8
+ */
+#define DUP_ENGINE_IO_REGION_0_OFFSET		f9cfc000
+#define DUP_ENGINE_IO_REGION_1_OFFSET		f9cfd000
+
+#define DUP_ENGINE_IO_REGION_0_PUSH_BASE	f9cf8000
+/* DUP_ENGINE_IO_REGION_0_PUSH_BASE + DUP_NUM_GROUPS_PER_REGION * DUP_GROUP_MMIO_RANGE) */
+#define DUP_ENGINE_IO_REGION_1_PUSH_BASE	f9cf9000
+
+#define DUP_ENGINE_COMPUTE_REGION_0_OFFSET	f1cfc000
+#define DUP_ENGINE_COMPUTE_REGION_1_OFFSET	f1cfd000
+
+#define DUP_ENGINE_COMPUTE_REGION_0_PUSH_BASE	f1cf8000
+/* DUP_ENGINE_COMPUTE_REGION_0_PUSH_BASE + DUP_NUM_GROUPS_PER_REGION * DUP_GROUP_MMIO_RANGE) */
+#define DUP_ENGINE_COMPUTE_REGION_1_PUSH_BASE	f1cf9000
+
+/* ARC mgmt dup group id start */
+#define ARC_MGMT_DUP_START_IDX			0
+#define ARC_MGMT_NUM_DUP_GROUPS			8
+
+#define ARC_MGMT_COMPUTE_DUP_REGION0_START_IDX		0
+#define ARC_MGMT_COMPUTE_DUP_REGION1_START_IDX		0
+#define ARC_MGMT_COMPUTE_NUM_DUP_GROUPS		8
+
+/* Arc pid dup group id start */
+#define ARC_PID_DUP_START_IDX		0
+#define ARC_PID_NUM_DUP_GROUPS		8
+
+#define RL_GROUP_ID_DEC_IO	ARC_PID_DUP_START_IDX
+#define RL_GROUP_ID_ENC_IO	(RL_GROUP_ID_DEC_IO + 1)
+#define RL_GROUP_ID_NIC_IO	(RL_GROUP_ID_ENC_IO + 1)
+
+#define RL_GROUP_ID_EDMA_COMPUTE	(ARC_MGMT_COMPUTE_DUP_REGION0_START_IDX)
+#define RL_GROUP_ID_PDMA_COMPUTE	(RL_GROUP_ID_EDMA_COMPUTE + 1)
+
+#define RL_GROUP_ID_CU0_RRTR_RL_GRP_0		(ARC_MGMT_COMPUTE_DUP_REGION1_START_IDX)
+#define RL_GROUP_ID_CU0_RRTR_RL_GRP_1		(RL_GROUP_ID_CU0_RRTR_RL_GRP_0 + 1)
+#define RL_GROUP_ID_CU0_RRTR_RL_GRP_2		(RL_GROUP_ID_CU0_RRTR_RL_GRP_1 + 1)
+#define RL_GROUP_ID_CU0_RRTR_RL_MSTR_GRP	(RL_GROUP_ID_CU0_RRTR_RL_GRP_2 + 1)
+
+#define RL_GROUP_ID_CU1_RRTR_RL_GRP_0		(RL_GROUP_ID_CU0_RRTR_RL_MSTR_GRP + 1)
+#define RL_GROUP_ID_CU1_RRTR_RL_GRP_1		(RL_GROUP_ID_CU1_RRTR_RL_GRP_0 + 1)
+#define RL_GROUP_ID_CU1_RRTR_RL_GRP_2		(RL_GROUP_ID_CU1_RRTR_RL_GRP_1 + 1)
+#define RL_GROUP_ID_CU1_RRTR_RL_MSTR_GRP	(RL_GROUP_ID_CU1_RRTR_RL_GRP_2 + 1)
+
+#define ARC_AGENT_ICCM_SECTION	1
+#define ARC_AGENT_ICCM_OFFSET	0
+#define ARC_AGENT_ICCM_HI	0x0300007F
+#define ARC_AGENT_ICCM_LO	0xF8840000
+#define ARC_AGENT_ICCM_SIZE	0x100
+#define ARC_AGENT_DCCM_SECTION	1
+#define ARC_AGENT_DCCM_OFFSET	80000000
+#define ARC_AGENT_DCCM_HI	0x0300007F
+#define ARC_AGENT_DCCM_LO	0xF8880000
+#define ARC_AGENT_DCCM_SIZE	0x80
+
+/* IO0 HALF 0 SW MBX */
+/* CMPT0 MGMT --> IO MGMT */
+#define IO0_HALF0_SW_MAILBOX0_0					F9CA7000
+/* CMPT1 MGMT --> IO MGMT */
+#define IO0_HALF0_SW_MAILBOX0_1					F9CA7004
+/* IO MGMT --> IO PID */
+#define IO0_HALF0_SW_MAILBOX1_0					F9CA8000
+/* IO MGMT --> IO RESET */
+#define IO0_HALF0_SW_MAILBOX1_1					F9CA8004
+/* IO PID --> IO MGMT */
+#define IO0_HALF0_SW_MAILBOX2_0					F9CA9000
+/* IO RESET --> IO MGMT */
+#define IO0_HALF0_SW_MAILBOX2_1					F9CA9004
+
+/* COMPUTE0 SW MBX */
+/* IO MGMT --> CMPT0 MGMT */
+#define COMPUTE0_SW_MAILBOX0_0					F1CA7000
+/* CMPT0 RESET --> CMPT0 MGMT */
+#define COMPUTE0_SW_MAILBOX1_0					F1CA8000
+/* CMPT0 MGMT --> CMPT0 RESET */
+#define COMPUTE0_SW_MAILBOX1_1					F1CA8004
+
+/* COMPUTE1 SW MBX */
+/* IO MGMT --> CMPT1 MGMT */
+#define COMPUTE1_SW_MAILBOX0_0					F3CA7000
+/* CMPT1 RESET --> CMPT0 MGMT */
+#define COMPUTE1_SW_MAILBOX1_0					F3CA8000
+/* CMPT1 MGMT --> CMPT0 RESET */
+#define COMPUTE1_SW_MAILBOX1_1					F3CA8004
+
+#endif
diff --git a/fs/jgs/jgs_reg_map.h b/fs/jgs/jgs_reg_map.h
new file mode 100644
index 0000000..ca52767
--- /dev/null
+++ b/fs/jgs/jgs_reg_map.h
@@ -0,0 +1,155 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+ #ifndef JGS_REG_MAP_H_
+ #define JGS_REG_MAP_H_
+
+ #define IO0_H0_PSOC_GLOBAL_CONF(reg)	(mmIO_0_CFG_PSOC_PSOC_CFG_GLOBAL_CONF_BASE + (reg))
+ #define IO0_H0_PSOC_BOOT(reg)		(mmIO_0_CFG_PSOC_BOOT_BOOT_BASE + (reg))
+ #define IO0_H0_PCIE_WRAP(reg)		(mmIO_0_CFG_CXL_EP_PCIE_WRAP_BASE + (reg))
+ #define IO0_H0_PCIE_DBI(reg)		(mmIO_0_CFG_CXL_EP_DBI_SIG_BASE + (reg))
+ #define IO0_H0_PCIE_ELBI(reg)		(mmIO_0_CFG_CXL_EP_PCIE_ELBI_BASE + (reg))
+ #define IO0_H0_PCIE_CORE(reg)		(mmIO_0_CFG_CXL_EP_PCIE_CORE_BASE + (reg))
+ #define IO0_H0_PCIE_FLR(reg)		(mmIO_0_CFG_CXL_EP_PCIE_FLR_BASE + (reg))
+ #define IO0_H0_PCIE_AUX(reg)		(mmIO_0_CFG_CXL_EP_PCIE_AUX_BASE + (reg))
+ #define IO0_H0_PCIE_PIF_ARC_WRAP(reg)	(mmIO_0_CFG_CXL_EP_PIF_ARC_WRAP_BASE + (reg))
+ #define IO0_H0_MNG_ARC(reg)		(mmIO_0_CFG_PSOC_ARC_MNG_ARC_BASE + (reg))
+ #define IO0_H0_BMU_CFG(reg)		(mmIO_0_CFG_CXL_EP_PCIE_BMU_CFG_BASE + (reg))
+ #define IO0_H0_BMU_MEM(reg)		(mmIO_0_CFG_CXL_EP_PCIE_BMU_MEM_BASE + (reg))
+ #define mmIO0_H0_PCIE_IDE(reg)		(mmIO_0_CFG_PCIE_EP_IDE_BASE + (reg))
+ #define mmIO0_H0_PCIE_IDE_AES(reg)	(mmIO_0_CFG_PCIE_EP_IDE_PCIE_IDE_AES_BASE + (reg))
+ #define mmIO0_H0_PCIE_IDE_STATE(reg)	(mmIO_0_CFG_PCIE_EP_IDE_PCIE_IDE_STATE_BASE + (reg))
+ #define IO0_H0_PMMU_PCIE_WRAP(reg)	(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_PMMU_PMMU_PCIE_WRAP_BASE + (reg))
+
+ #define IO0_H0_PSOC_ARC_MNG_LOCAL(reg)	(mmIO_0_CFG_PSOC_ARC_MNG_LOCAL_BASE + (reg))
+ #define IO0_H0_PSOC_BOOT_SECURITY(reg)	(mmIO_0_CFG_PSOC_BOOT_SECURITY_BASE + (reg))
+ #define IO0_H0_PSOC_TIMESTAMP(reg)	(mmIO_0_CFG_PSOC_PSOC_CFG_TIMESTAMP_BASE + (reg))
+ #define IO0_H0_PSOC_QSPI(reg)		(mmIO_0_CFG_PSOC_PSOC_CFG_QSPI_BASE + (reg))
+ #define IO0_H0_PSOC_SPI(reg)		(mmIO_0_CFG_PSOC_PSOC_CFG_SPI_BASE + (reg))
+ #define IO0_H0_PSOC_D2D_SPI(reg)	(mmIO_0_CFG_PSOC_PSOC_CFG_D2D_SPI_BASE + (reg))
+ #define IO0_H0_PSOC_EFUSE(reg)		(mmIO_0_CFG_PSOC_PSOC_CFG_EFUSE_BASE + (reg))
+ #define I0_H0_PSOC_GPIO0(reg)          (mmIO_0_CFG_PSOC_GPIO0_BASE + (reg))
+ #define I0_H0_PSOC_GPIO1(reg)          (mmIO_0_CFG_PSOC_GPIO1_BASE + (reg))
+ #define I0_H0_PSOC_GPIO4(reg)          (mmIO_0_CFG_PSOC_GPIO4_BASE + (reg))
+ #define IO0_H0_PSOC_PAD_CONF(reg)	(mmIO_0_CFG_PSOC_PSOC_CFG_PAD_CONF_BASE + (reg))
+
+ #define IO0_H0_SOC_0_ID2D_PHY_0_CTL0_BASE(reg)	\
+			 (mmIO_0_CFG_SOC_0_ID2D_PHY_0_D2D_CTL0_BASE + (reg))
+ #define IO0_H0_SOC_0_ID2D_PHY_1_CTL0_BASE(reg)	\
+			 (mmIO_0_CFG_SOC_0_ID2D_PHY_1_D2D_CTL0_BASE + (reg))
+ #define IO0_H1_SOC_1_ID2D_PHY_0_CTL0_BASE(reg)	\
+			 (mmIO_0_CFG_SOC_1_ID2D_PHY_0_D2D_CTL0_BASE + (reg))
+ #define IO0_H1_SOC_1_ID2D_PHY_1_CTL0_BASE(reg)	\
+			 (mmIO_0_CFG_SOC_1_ID2D_PHY_1_D2D_CTL0_BASE + (reg))
+ #define IO0_H0_SOC_0_ID2D_MAC_0_D2D_MAC0_BASE(reg)  \
+			 (mmIO_0_CFG_SOC_0_ID2D_MAC_0_D2D_MAC0_BASE + (reg))
+ #define IO0_H0_SOC_0_ID2D_MAC_1_D2D_MAC0_BASE(reg)  \
+			 (mmIO_0_CFG_SOC_0_ID2D_MAC_1_D2D_MAC0_BASE + (reg))
+ #define IO0_H1_SOC_1_ID2D_MAC_0_D2D_MAC0_BASE(reg)  \
+			 (mmIO_0_CFG_SOC_1_ID2D_MAC_0_D2D_MAC0_BASE + (reg))
+ #define IO0_H1_SOC_1_ID2D_MAC_1_D2D_MAC0_BASE(reg)  \
+			 (mmIO_0_CFG_SOC_1_ID2D_MAC_1_D2D_MAC0_BASE + (reg))
+ #define IO0_H0_ID2D_PHY_0_D2D_PHY0_CH_BCAST_PHYCTRLREG_BASE(reg)	\
+			 mmIO_0_CFG_SOC_0_ID2D_PHY_0_D2D_PHY0_CH_BCAST_PHYCTRLREG_BASE + (reg)
+ #define IO0_H0_PCIE_LBW_MSTR_IF_DBG(reg)	\
+			 (mmIO_0_CFG_CXL_EP_PCIE_LBW_MSTR_IF_PF_LBW_DBG_BASE + reg)
+
+ #define COMP_0_CFG_COMP_SOC_VD2D_PHY_0_COMP_DPHY_CTRL_BASE(reg)	\
+			 mmCOMPUTE_0_CFG_COMP_SOC_VD2D_PHY_0_COMP_DPHY_CTRL_BASE + (reg)
+ #define COMP_0_CFG_COMP_SOC_VD2D_PHY_0_IPT_D2D_N3_TOP_CH_BCAST_PHYCTRLREG_BASE(reg)	\
+	 mmCOMPUTE_0_CFG_COMP_SOC_VD2D_PHY_0_IPT_D2D_N3_TOP_CH_BCAST_PHYCTRLREG_BASE + (reg)
+ #define COMP_0_CFG_PSOC_BOOT_RESET_ISO(reg)	(mmCOMPUTE_0_CFG_PSOC_BOOT_RESET_ISO_BASE + (reg))
+
+ #define IO0_H0_PSOC_CFG_UART(reg)	(mmIO_0_CFG_PSOC_PSOC_CFG_UART_BASE + (reg))
+ #define IO0_H0_PSOC_CFG_PAD_CONF_PAD_SEL(reg)	(mmIO_0_CFG_PSOC_PSOC_CFG_PAD_CONF_BASE + (reg))
+ #define IO0_H0_SOC_0_PLL_0_PSOC_PLL(reg)	\
+ 			 (mmIO_0_CFG_PLL_GRP0_PSOC_PLL_PLL_CTRL_BASE + (reg))
+ #define ID0_H0_PSOC_BOOT_RESET_ISO(reg)		(mmIO_0_CFG_PSOC_BOOT_RESET_ISO_BASE + (reg))
+ #define IO0_H0_PSOC_PARC_CFG_MISC_BASE(reg)	(mmIO_0_CFG_PSOC_PARC_CFG_MISC_BASE + (reg))
+
+ #define IO0_H0_PSOC_CFG_MEM_COPY(reg)	(mmIO_0_CFG_PSOC_PSOC_CFG_MEM_COPY_BASE + (reg))
+ #define IO0_H0_PSOC_CFG_MEM_COPY_LBW_TRAN_CHK(reg)	\
+		 (mmIO_0_CFG_PSOC_PSOC_CFG_MEM_COPY_MSTR_IF_LBW_TSRL_T_BASE + (reg))
+
+ #define mmCPU_BOOT_DEV_STS0		IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_DEV_STS0)
+ #define mmCPU_BOOT_DEV_STS1		IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_DEV_STS1)
+ #define mmCPU_CMD_STATUS_TO_HOST	IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_CMD_STATUS_TO_HOST)
+ #define mmCPU_BOOT_ERR0			IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_ERR0)
+ #define mmCPU_BOOT_ERR1			IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_CPU_BOOT_ERR1)
+ #define mmUPD_STS			IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_UPD_STS)
+ #define mmUPD_CMD			IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_UPD_CMD)
+
+ #define mmUPD_PENDING_STS		IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_UPD_PENDING)
+ #define mmCOLD_RST_DATA			IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_DATA)
+ #define mmHW_STATE			IO0_H0_MNG_ARC(mmARC_KMD_HW_DIRTY_STATUS)
+
+ #define mmVUART_MNG0_TX_DATA		IO0_H0_PMMU_PCIE_WRAP(mmPMMU_PCIE_WRAP_VUART_TX_0)
+ #define mmVUART_MNG0_TX_STATUS		IO0_H0_PMMU_PCIE_WRAP(mmPMMU_PCIE_WRAP_VUART_TX_1)
+ #define mmVUART_MNG0_RX_DATA		IO0_H0_PMMU_PCIE_WRAP(mmPMMU_PCIE_WRAP_VUART_RX_0)
+ #define mmVUART_MNG0_RX_STATUS		IO0_H0_PMMU_PCIE_WRAP(mmPMMU_PCIE_WRAP_VUART_RX_1)
+
+ /*
+  * NOTE: All register offsets used for regmap_readl & regmap_writel should be of
+  * IO DIE0 as the api will perform address manipulation for other dies.
+  * Hardware performs address manipulation only for PARC_LOCAL_CFG page in configuration
+  * space where DIE ID register is available.
+  */
+ #define FW_BLOCK_BASE(SPECS_BLOCK_BASE)	(SPECS_BLOCK_BASE & 0xFFFFFFFF)
+
+ #define mmPARC_LOCAL_DIE_ID		(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_ARC_MNG_LOCAL_BASE) \
+					+ mmARC_LOCAL_CONF_DIE_CPU_ID)
+ #define mmI2C_S_BASE			(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_I2C_S_BASE))
+ #define mmI2C_S_IC_CON			(mmI2C_S_BASE + mmI2C_IC_CON)
+ #define mmI2C_S_IC_SAR			(mmI2C_S_BASE + mmI2C_IC_SAR)
+ #define mmI2C_S_IC_INTR_MASK		(mmI2C_S_BASE + mmI2C_IC_INTR_MASK)
+ #define mmI2C_S_IC_RX_TL		(mmI2C_S_BASE + mmI2C_IC_RX_TL)
+ #define mmI2C_S_IC_TX_TL		(mmI2C_S_BASE + mmI2C_IC_TX_TL)
+ #define mmI2C_S_IC_ENABLE		(mmI2C_S_BASE + mmI2C_IC_ENABLE)
+
+ #define mmQSPI_BASE			(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_QSPI_BASE))
+ #define mmQSPI_SER			(mmQSPI_BASE + mmSPI_SER)
+ #define mmQSPI_BAUDR			(mmQSPI_BASE + mmSPI_BAUDR)
+ #define mmQSPI_DR0			(mmQSPI_BASE + mmSPI_DR0)
+ #define mmSPI_WRITE_LIMIT		(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_GLOBAL_CONF_BASE) \
+					 + mmGLOBAL_CONF_SPI_WRITE_LIMIT)
+
+ #define mmTIMESTAMP_BASE        (FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_TIMESTAMP_BASE))
+ #define mmIO_0_CFG_HALF_0_TIMESTAMP_CNTCR	(mmTIMESTAMP_BASE + mmTIMESTAMP_CNTCR)
+ #define mmIO_0_CFG_HALF_0_TIMESTAMP_CNTSR	(mmTIMESTAMP_BASE + mmTIMESTAMP_CNTSR)
+ #define mmIO_0_CFG_HALF_0_TIMESTAMP_CNTCVL	(mmTIMESTAMP_BASE + mmTIMESTAMP_CNTCVL)
+ #define mmIO_0_CFG_HALF_0_TIMESTAMP_CNTCVU	(mmTIMESTAMP_BASE + mmTIMESTAMP_CNTCVU)
+ #define mmIO_0_CFG_HALF_0_TIMESTAMP_CNTFID0	(mmTIMESTAMP_BASE + mmTIMESTAMP_CNTFID0)
+
+ #define mmPRSTN                 (FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_GLOBAL_CONF_BASE) \
+				 + mmGLOBAL_CONF_PRSTN)
+ #define mmCOLD_RST_FLOPS_2	(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_BOOT_BOOT_BASE)\
+				 + mmPSOC_BOOT_CONF_COLD_RST_REG_2)
+ #define mmCOLD_RST_FLOPS_3	(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_BOOT_BOOT_BASE)\
+				 + mmPSOC_BOOT_CONF_COLD_RST_REG_3)
+ #define mmPCIE_PHY_RST_N	(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_GLOBAL_CONF_BASE) \
+				 + mmGLOBAL_CONF_PCIE_PHY_RST_N)
+ #define mmBOOT_STRAP_PINS_L	(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_BOOT_BOOT_BASE) \
+				 + mmPSOC_BOOT_CONF_BOOT_STRAP_PINS_L)
+ #define mmPRSTN_MASK		(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_BOOT_BOOT_BASE) \
+				  + mmPSOC_BOOT_CONF_PRSTN_RST_MASK)
+
+
+ #define mmSPI_DIRECT_WR_RD_CTRL		(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_GLOBAL_CONF_BASE) \
+					 + mmGLOBAL_CONF_SPI_DIRECT_WR_RD_CTRL)
+ #define mmSYSCFG1_EFUSE_DATA		(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_GLOBAL_CONF_BASE) \
+					 + mmGLOBAL_CONF_SCRATCHPAD_0)
+ #define mmSYSCFG2_EFUSE_DATA		(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_GLOBAL_CONF_BASE) \
+					 + mmGLOBAL_CONF_SCRATCHPAD_1)
+ #define mmFFUSE_WRITE_IND		(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_PSOC_CFG_GLOBAL_CONF_BASE) \
+					 + mmGLOBAL_CONF_SCRATCHPAD_2)
+
+ /* CPLD watchdog GPIO is part of GPIO4 */
+ #define mmGPIO4_DATA			(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_GPIO4_BASE) \
+					 + mmGPIO_DATA)
+ #define mmGPIO4_OUTENSET		(FW_BLOCK_BASE(mmIO_0_CFG_PSOC_GPIO4_BASE) \
+					 + mmGPIO_OUTENSET)
+
+ #endif /* JGS_REG_MAP_H_ */
diff --git a/fs/jgs/jgs_reg_mask.h b/fs/jgs/jgs_reg_mask.h
new file mode 100644
index 0000000..4574035
--- /dev/null
+++ b/fs/jgs/jgs_reg_mask.h
@@ -0,0 +1,51 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+ #ifndef _JGS_REG_MASK_H_
+ #define _JGS_REG_MASK_H_
+
+ #define SPI_DMA_CTRL_ADDR_SHIFT			PARC_GLOBAL_CONF_SPI_DMA_CTRL_ADDR_SHIFT_S
+ #define SPI_DMA_CTRL_DST_SRAM_SHIFT		PARC_GLOBAL_CONF_SPI_DMA_CTRL_DST_SRAM_S
+ #define SPI_DMA_STATUS_ERROR_MASK		PARC_GLOBAL_CONF_SPI_DMA_STATUS_ERROR_M
+ #define SPI_DMA_STATUS_ERROR_SHIFT		PARC_GLOBAL_CONF_SPI_DMA_STATUS_ERROR_S
+ #define SPI_DMA_STATUS_DONE_MASK		PARC_GLOBAL_CONF_SPI_DMA_STATUS_DONE_M
+ #define SPI_DMA_CTRL_FLD_SIZE_SHIFT		PARC_GLOBAL_CONF_SPI_DMA_CTRL_MEM_SIZE_S
+ #define SPI_DIRECT_WR_RD_CTRL_BYTE_SWAP_MASK	\
+					 GLOBAL_CONF_SPI_DIRECT_WR_RD_CTRL_BYTE_SWAP_M
+ #define PRSTN_MASK_IND_MASK			PSOC_BOOT_CONF_PRSTN_RST_MASK_IND_M
+ #define MEM_CPY_CTRL2_USE_CONST_SHIFT		PSOC_CFG_GLOBAL_CONF_MEM_CPY_CTRL2_USE_CONST_S
+
+ #define SECURITY_PRI_IMG_STATUS_FW_VER_LOWER_SHIFT	\
+					 PSOC_SECURITY_PRI_IMG_STATUS_FW_VER_LOWER_S
+ #define SECURITY_PRI_IMG_STATUS_AUTHENTICATE_FAIL_SHIFT	\
+					 PSOC_SECURITY_PRI_IMG_STATUS_AUTHENTICATE_FAIL_S
+ #define SECURITY_PRI_IMG_STATUS_CRC_FAIL_SHIFT	PSOC_SECURITY_PRI_IMG_STATUS_CRC_FAIL_S
+ #define SECURITY_STATUS_IMG_VALIDATED_SHIFT	PSOC_SECURITY_STATUS_IMG_VALIDATED_S
+ #define SECURITY_STATUS_FW_UPGRADE_SHIFT	PSOC_SECURITY_STATUS_FW_UPGRADE_S
+ #define SECURITY_STATUS_KEY_REVOKED_SHIFT	PSOC_SECURITY_STATUS_KEY_REVOKED_S
+ #define SECURITY_STATUS_KEY_REVOKED_IND_SHIFT	PSOC_SECURITY_STATUS_KEY_REVOKED_IND_S
+ #define SECURITY_STATUS_MIN_FW_MODIFIED_SHIFT	PSOC_SECURITY_STATUS_MIN_FW_MODIFIED_S
+ #define SECURITY_STATUS_FW_VER_INC_SHIFT	PSOC_SECURITY_STATUS_FW_VER_INC_S
+ #define SECURITY_STATUS_SPI_IMG_MODIFIES_SHIFT	PSOC_SECURITY_STATUS_SPI_IMG_MODIFIES_S
+ #define SECURITY_STATUS_FW_AUTHENTICAION_FAIL_SHIFT	\
+					 PSOC_SECURITY_STATUS_FW_AUTHENTICAION_FAIL_S
+
+ #define LINK_CAPABILITIES_REG_PCIE_CAP_MAX_LINK_SPEED_MASK	\
+				 PCIE_DBI_LINK_CAPABILITIES_REG_PCIE_CAP_MAX_LINK_SPEED_M
+ #define LINK_CAPABILITIES_REG_PCIE_CAP_MAX_LINK_SPEED_SHIFT	\
+				 PCIE_DBI_LINK_CAPABILITIES_REG_PCIE_CAP_MAX_LINK_SPEED_S
+ #define LINK_CONTROL2_LINK_STATUS2_REG_PCIE_CAP_TARGET_LINK_SPEED_MASK	\
+	 PCIE_DBI_LINK_CONTROL2_LINK_STATUS2_REG_PCIE_CAP_TARGET_LINK_SPEED_M
+ #define LINK_CONTROL2_LINK_STATUS2_REG_PCIE_CAP_TARGET_LINK_SPEED_SHIFT	\
+	 PCIE_DBI_LINK_CONTROL2_LINK_STATUS2_REG_PCIE_CAP_TARGET_LINK_SPEED_S
+
+ #define LTSSM_EN_IND_MASK			PCIE_AUX_LTSSM_EN_IND_M
+
+ #define CPU_HOT_RST_CTRL_MASK			PCIE_CORE_HOT_RST_INTR_M
+
+ #define FORCE_RECOVERY_MASK			BIT(1)	//GPIO_01
+
+ #endif /* _JGS_REG_MASK_H_ */
diff --git a/fs/jgs/jgs_security.h b/fs/jgs/jgs_security.h
new file mode 100644
index 0000000..3d5bd82
--- /dev/null
+++ b/fs/jgs/jgs_security.h
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2023 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+ #ifndef _JGS_SECURITY_H_
+ #define _JGS_SECURITY_H_
+
+ #define HL_SEC_FW_VERSION			0
+ #define HL_SEC_KEY_INDEX			0x0
+
+ enum hl_sec_tpm_pcr_index {
+	 TPM2_SVN_PCR_IDX = 0,
+	 TPM2_MANIFEST_PCR_IDX = 1,
+	 TPM2_BOOT_FIT_PCR_IDX = 20,
+ };
+
+ #endif /* _JGS_SECURITY_H_ */
